File: Python/Partie7/Exo30.py
import signal

def segfault_handler(sig, frame):
    print("Erreur de segmentation, réessayez")
    raise Exception("segfault")

signal.signal(signal.SIGSEGV, segfault_handler)

n = int(input("Nombre d'éléments : "))
tab = [0]*n
for i in range(n):
    tab[i] = int(input(f"tab[{i}] = "))
while True:
    i = int(input("Indice à afficher : "))
    try:
        print(f"tab[{i}] = {tab[i]}")
        break
    except Exception:
        print("Erreur de segmentation, réessayez")

File: C/Partie7/Exo30.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <setjmp.h>

jmp_buf env;

void segfault_handler(int sig)
{
    longjmp(env, 1);
}

int main()
{
    signal(SIGSEGV, segfault_handler);
    int n;
    printf("Nombre d'éléments : ");
    scanf("%d", &n);
    int *tab = malloc(n * sizeof(int));
    printf("Entrez %d entiers :\n", n);
    for (int i = 0; i < n; i++)
        scanf("%d", &tab[i]);
    int i;
    while (1)
    {
        printf("Indice à afficher : ");
        scanf("%d", &i);
        if (setjmp(env) == 0)
        {
            printf("tab[%d] = %d\n", i, tab[i]);
            break;
        }
        else
        {
            printf("Erreur de segmentation, réessayez\n");
        }
    }
    free(tab);
    return 0;
}